cuda/matrix/csr_kernels.cu

```diff
@@ -47,36 +53,518 @@ namespace csr {
+                static_cast<IndexType>(a->get_size()[0]),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                as_cuda_type(a->get_const_row_ptrs()),
+                as_cuda_type(a->get_const_srow()),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(c->get_values()), c->get_stride(),
+                as_cuda_type(row_out.get_data()),
+                as_cuda_type(val_out.get_data()));
+        } else if (num_item == 14) {
+            merge_path_spmv<14 / index_scale><<<grid, block>>>(
+                static_cast<IndexType>(a->get_size()[0]),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                as_cuda_type(a->get_const_row_ptrs()),
+                as_cuda_type(a->get_const_srow()),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(c->get_values()), c->get_stride(),
+                as_cuda_type(row_out.get_data()),
+                as_cuda_type(val_out.get_data()));
+        }
+
+        reduce<<<1, spmv_block_size>>>(
+            grid_num, as_cuda_type(val_out.get_data()),
+            as_cuda_type(row_out.get_data()), as_cuda_type(c->get_values()),
+            c->get_stride());
+    } else if (a->get_strategy()->get_name() == "classical") {
+        classical_spmv<<<ceildiv(a->get_size()[0], classical_block_size),
+                         classical_block_size>>>(
+            a->get_size()[0], as_cuda_type(a->get_const_values()),
+            a->get_const_col_idxs(), as_cuda_type(a->get_const_row_ptrs()),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+    } else if (a->get_strategy()->get_name() == "cusparse") {
+        if (cusparse::is_supported<ValueType, IndexType>::value) {
+            // TODO: add implementation for int64 and multiple RHS
+            auto handle = cusparse::init();
+            auto descr = cusparse::create_mat_descr();
+            ASSERT_NO_CUSPARSE_ERRORS(
+                cusparseSetPointerMode(handle, CUSPARSE_POINTER_MODE_HOST));
+
+            auto row_ptrs = a->get_const_row_ptrs();
+            auto col_idxs = a->get_const_col_idxs();
+            auto alpha = one<ValueType>();
+            auto beta = zero<ValueType>();
+            if (b->get_stride() != 1 || c->get_stride() != 1) NOT_IMPLEMENTED;
+
+            cusparse::spmv(handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                           a->get_size()[0], a->get_size()[1],
+                           a->get_num_stored_elements(), &alpha, descr,
+                           a->get_const_values(), row_ptrs, col_idxs,
+                           b->get_const_values(), &beta, c->get_values());
+
+            cusparse::destroy(descr);
+            cusparse::destroy(handle);
+        } else {
+            NOT_IMPLEMENTED;
```

Comment:
Looks good. just a final minor thing: instead of `NOT_IMPLEMENTED` we could fall back to classical SpMV here.

Reply:
done